pkg/ingester/instance.go

```diff
@@ -180,80 +178,132 @@ func (i *instance) Push(ctx context.Context, req *logproto.PushRequest) error {
-		if record != nil {
-			record.Series = append(record.Series, tsdb_record.RefSeries{
-				Ref:    chunks.HeadSeriesRef(fp),
-				Labels: sortedLabels,
-			})
-		} else {
-			// If the record is nil, this is a WAL recovery.
-			i.metrics.recoveredStreamsTotal.Inc()
+		validation.DiscardedSamples.WithLabelValues(validation.StreamLimit, i.instanceID).Add(float64(len(pushReqStream.Entries)))
+		bytes := 0
+		for _, e := range pushReqStream.Entries {
+			bytes += len(e.Line)
 		}
+		validation.DiscardedBytes.WithLabelValues(validation.StreamLimit, i.instanceID).Add(float64(bytes))
+		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, validation.StreamLimitErrorMsg)
+	}
 
-		memoryStreams.WithLabelValues(i.instanceID).Inc()
-		i.streamsCreatedTotal.Inc()
-		i.addTailersToNewStream(s)
-
+	labels, err := logql.ParseLabels(pushReqStream.Labels)
+	if err != nil {
 		if i.configs.LogStreamCreation(i.instanceID) {
 			level.Debug(util_log.Logger).Log(
-				"msg", "successfully created stream",
+				"msg", "failed to create stream, failed to parse labels",
 				"org_id", i.instanceID,
+				"err", err,
 				"stream", pushReqStream.Labels,
 			)
 		}
-		return s, nil
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+	fp := i.getHashForLabels(labels)
+
+	sortedLabels := i.index.Add(cortexpb.FromLabelsToLabelAdapters(labels), fp)
+	s := newStream(i.cfg, i.limiter, i.instanceID, fp, sortedLabels, i.limiter.UnorderedWrites(i.instanceID), i.metrics)
+
+	// record will be nil when replaying the wal (we don't want to rewrite wal entries as we replay them).
+	if record != nil {
+		record.Series = append(record.Series, tsdb_record.RefSeries{
+			Ref:    chunks.HeadSeriesRef(fp),
+			Labels: sortedLabels,
+		})
+	} else {
+		// If the record is nil, this is a WAL recovery.
+		i.metrics.recoveredStreamsTotal.Inc()
+	}
+
+	memoryStreams.WithLabelValues(i.instanceID).Inc()
+	i.streamsCreatedTotal.Inc()
+	i.addTailersToNewStream(s)
+
+	if i.configs.LogStreamCreation(i.instanceID) {
+		level.Debug(util_log.Logger).Log(
+			"msg", "successfully created stream",
+			"org_id", i.instanceID,
+			"stream", pushReqStream.Labels,
+		)
+	}
+
+	return s, nil
+}
+
+// getOrCreateStream returns the stream or creates it.
+func (i *instance) getOrCreateStream(pushReqStream logproto.Stream, record *WALRecord) (*stream, error) {
+	s, _, err := i.streams.LoadOrStoreNew(pushReqStream.Labels, func() (*stream, error) {
+		return i.createStream(pushReqStream, record)
 	})
 
 	return s, err
 }
 
+// executeWithGetOrCreateStream executes fn after find or create stream, with chunkMtx locked.
+// Use this to keep Load and Delete consistency for streams while executing fn.
+// If fn doesn't care about stream deletion from streams, use getOrCreateStream directly(e.g. ingesterRecoverer).
+func (i *instance) executeWithGetOrCreateStream(pushReqStream logproto.Stream, record *WALRecord, fn func(*stream) error) error {
```

Comment:
It may be more intuitive to provide a helper function which will get/create a stream and execute a function under one lock acquisition. In addition to reducing the streams lock contention by 50% (no need to double check after the first create), it will present a simpler interface. Something like:
```
// Executes a function on a stream with the lock held.
// Will create the stream according to `newStreamFn` if it doesn't exist
func (m *streamsMap) UnderLock(
	pushReqStream logproto.Stream,
	newStreamFn func() (*stream, error),
	fn func(*stream) error,
	) (error) {
	// lock
	// get or create
	// run fn on stream
	// unlock
}
```

Reply:
The lock here is `chunkMtx` under `stream`, instead of `consistencyMtx` under `streamsMap`. Which means we have to get the stream first, then lock the `chunkMtx` under it. The reason to lock `chunkMtx` is to make sure `stream` is not being deleted by `removeFlushedChunks` while `fn` is running.